File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/TableEndpointDefinition.java
package com.raizlabs.android.dbflow.processor.definition;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.raizlabs.android.dbflow.annotation.provider.ContentUri;
import com.raizlabs.android.dbflow.annotation.provider.Notify;
import com.raizlabs.android.dbflow.annotation.provider.TableEndpoint;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;

import java.util.List;
import java.util.Map;

import javax.lang.model.element.Element;
import javax.lang.model.element.PackageElement;
import javax.lang.model.element.TypeElement;

/**
 * Description:
 */
public class TableEndpointDefinition extends BaseDefinition {

    public List<ContentUriDefinition> contentUriDefinitions = Lists.newArrayList();

    /**
     * Dont want duplicate paths.
     */
    Map<String, ContentUriDefinition> pathValidationMap = Maps.newHashMap();

    public Map<String, Map<Notify.Method, List<NotifyDefinition>>> notifyDefinitionPathMap = Maps.newHashMap();

    public String tableName;

    public String contentProviderName;

    public boolean isTopLevel = false;

    public TableEndpointDefinition(Element typeElement, ProcessorManager processorManager) {
        super(typeElement, processorManager);

        TableEndpoint endpoint = typeElement.getAnnotation(TableEndpoint.class);

        tableName = endpoint.name();

        contentProviderName = endpoint.contentProviderName();

        isTopLevel = typeElement.getEnclosingElement() instanceof PackageElement;

        List<? extends Element> elements = processorManager.getElements().getAllMembers((TypeElement) typeElement);
        for (Element innerElement : elements) {
            if (innerElement.getAnnotation(ContentUri.class) != null) {
                ContentUriDefinition contentUriDefinition = new ContentUriDefinition(innerElement, processorManager);
                if (!pathValidationMap.containsKey(contentUriDefinition.path)) {
                    contentUriDefinitions.add(contentUriDefinition);
                } else {
                    processorManager.logError("There must be unique paths for the specified @ContentUri" +
                            " %1s from %1s", contentUriDefinition.name, contentProviderName);
                }
            } else if (innerElement.getAnnotation(Notify.class) != null) {
                NotifyDefinition notifyDefinition = new NotifyDefinition(innerElement, processorManager);

                for (String path : notifyDefinition.paths) {
                    Map<Notify.Method, List<NotifyDefinition>> methodListMap = notifyDefinitionPathMap.get(path);
                    if (methodListMap == null) {
                        methodListMap = Maps.newHashMap();
                        notifyDefinitionPathMap.put(path, methodListMap);
                    }

                    List<NotifyDefinition> notifyDefinitionList = methodListMap.get(notifyDefinition.method);
                    if (notifyDefinitionList == null) {
                        notifyDefinitionList = Lists.newArrayList();
                        methodListMap.put(notifyDefinition.method, notifyDefinitionList);
                    }
                    notifyDefinitionList.add(notifyDefinition);
                }
            }
        }

    }
}

File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/TableEndpointDefinition.kt
package com.raizlabs.android.dbflow.processor.definition

import com.raizlabs.android.dbflow.annotation.provider.ContentUri
import com.raizlabs.android.dbflow.annotation.provider.Notify
import com.raizlabs.android.dbflow.annotation.provider.TableEndpoint
import com.raizlabs.android.dbflow.processor.ProcessorManager
import com.squareup.javapoet.TypeName
import javax.lang.model.element.Element
import javax.lang.model.element.PackageElement
import javax.lang.model.element.TypeElement
import javax.lang.model.type.MirroredTypeException

/**
 * Description:
 */
class TableEndpointDefinition(typeElement: Element, processorManager: ProcessorManager)
: BaseDefinition(typeElement, processorManager) {

    var contentUriDefinitions: MutableList<ContentUriDefinition> = mutableListOf()

    /**
     * Dont want duplicate paths.
     */
    internal var pathValidationMap: Map<String, ContentUriDefinition> = mutableMapOf()

    var notifyDefinitionPathMap: MutableMap<String, MutableMap<Notify.Method, MutableList<NotifyDefinition>>>
            = mutableMapOf()

    var tableName: String? = null

    var contentProviderName: TypeName? = null

    var isTopLevel = false

    init {

        val endpoint = typeElement.getAnnotation(TableEndpoint::class.java)
        if (endpoint != null) {

            tableName = endpoint.name

            try {
                endpoint.contentProvider
            } catch (mte: MirroredTypeException) {
                contentProviderName = TypeName.get(mte.typeMirror)
            }

        }

        isTopLevel = typeElement.enclosingElement is PackageElement

        val elements = processorManager.elements.getAllMembers(typeElement as TypeElement)
        for (innerElement in elements) {
            if (innerElement.getAnnotation(ContentUri::class.java) != null) {
                val contentUriDefinition = ContentUriDefinition(innerElement, processorManager)
                if (!pathValidationMap.containsKey(contentUriDefinition.path)) {
                    contentUriDefinitions.add(contentUriDefinition)
                } else {
                    processorManager.logError("There must be unique paths for the specified @ContentUri" + " %1s from %1s", contentUriDefinition.name, contentProviderName)
                }
            } else if (innerElement.getAnnotation(Notify::class.java) != null) {
                val notifyDefinition = NotifyDefinition(innerElement, processorManager)

                for (path in notifyDefinition.paths) {
                    var methodListMap = notifyDefinitionPathMap[path]
                    if (methodListMap == null) {
                        methodListMap = mutableMapOf()
                        notifyDefinitionPathMap.put(path, methodListMap)
                    }

                    var notifyDefinitionList: MutableList<NotifyDefinition>? = methodListMap[notifyDefinition.method]
                    if (notifyDefinitionList == null) {
                        notifyDefinitionList = arrayListOf()
                        methodListMap.put(notifyDefinition.method, notifyDefinitionList)
                    }
                    notifyDefinitionList!!.add(notifyDefinition)
                }
            }
        }

    }
}
